analyzers/src/SonarAnalyzer.Common/Helpers/KnownMethods.cs

```diff
@@ -87,26 +87,17 @@ public static bool IsStaticObjectEquals(this IMethodSymbol methodSymbol)
         && methodSymbol.Parameters.Length == 0
         && methodSymbol.ReturnType.Is(KnownType.System_String);
 
-    public static bool IsIDisposableDispose(this IMethodSymbol methodSymbol)
-    {
-        const string explicitName = "System.IDisposable.Dispose";
-        return methodSymbol != null
-            && (methodSymbol.Name == nameof(IDisposable.Dispose) || methodSymbol.Name == explicitName)
-            && methodSymbol.Arity == 0
-            && methodSymbol.ReturnsVoid
-            && methodSymbol.Parameters.Length == 0;
-    }
-
-    public static bool IsIAsyncDisposableDisposeAsync(this IMethodSymbol methodSymbol)
-    {
-        const string explicitNameAsync = "System.IAsyncDisposable.DisposeAsync";
-        return methodSymbol != null
-            && (methodSymbol.Name == "DisposeAsync" || methodSymbol.Name == explicitNameAsync)
-            && methodSymbol.ReturnType.Is(KnownType.System_Threading_Tasks_ValueTask)
-            && methodSymbol.Parameters.Length == 0
-            && ContainingInterface(methodSymbol) is { } containingInterface
-            && containingInterface.Is(KnownType.System_IAsyncDisposable);
-    }
+    public static bool IsIDisposableDispose(this IMethodSymbol methodSymbol) =>
+        // ref struct types can also use the Disposable Pattern: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/using#pattern-based-using
+        methodSymbol is { IsStatic: false, Name: "Dispose" or "System.IDisposable.Dispose", Arity: 0, ReturnsVoid: true, Parameters.Length: 0 }
+        && ((ContainingInterface(methodSymbol) is { } containingInterface && containingInterface.Is(KnownType.System_IDisposable))
+            || (methodSymbol.ContainingType is { IsValueType: true } && methodSymbol.ContainingType.IsRefLikeType()));
```

Comment:
IMO nested conditionals where you need parentheses to declare precedence are not very readable.
I would split it in 3 parts like this:

```csharp
public static bool IsIDisposableDispose(this IMethodSymbol methodSymbol)
{
    var isDisposable = methodSymbol is
    {
        IsStatic: false,
        Name: "Dispose" or "System.IDisposable.Dispose",
        Arity: 0,
        ReturnsVoid: true,
        Parameters.Length: 0
    };
    var fromInterface = ContainingInterface(methodSymbol) is { } @interface && @interface.Is(KnownType.System_IDisposable);
    // ref struct types can also use the Disposable Pattern:
    // https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/using#pattern-based-using
    var fromRefStruct = methodSymbol.ContainingType?.IsValueType == true && methodSymbol.ContainingType.IsRefLikeType();
    return isDisposable && (fromInterface || fromRefStruct);
}
```

Reply:
I agree, it's more readable. But I prefer short-circuiting boolean operations for performance reasons. So I added more comments and refactored it like this:

```cs
// The Dispose method is either coming from System.IDisposable for classes and records or declared manually for ref struct types:
// https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/using#pattern-based-using
public static bool IsIDisposableDispose(this IMethodSymbol methodSymbol) =>
    methodSymbol is
    {
        IsStatic: false,
        Name: "Dispose" or "System.IDisposable.Dispose",
        Arity: 0,
        ReturnsVoid: true,
        Parameters.Length: 0
    }
    && ((ContainingInterface(methodSymbol) is { } containingInterface && containingInterface.Is(KnownType.System_IDisposable))  // class/record implementing System.IDisposable
        || (methodSymbol.ContainingType is { IsValueType: true } && methodSymbol.ContainingType.IsRefLikeType()));              // or a ref struct type
```

Reply:
I would assume that it will be in-lined.
But this version is also fine.